File: artifacts/api-server/routers/estoque.py
"""Rotas de Estoque."""
import sys, os; sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from typing import Optional, List
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from database import get_db
from models import Produto, MovimentoEstoque
from schemas import MovimentoEstoqueInput, MovimentoEstoqueOut

router = APIRouter()


@router.post("/movimento", response_model=MovimentoEstoqueOut, status_code=201)
def movimentar_estoque(data: MovimentoEstoqueInput, db: Session = Depends(get_db)):
    produto = db.query(Produto).filter(Produto.id == data.produto_id).first()
    if not produto:
        raise HTTPException(status_code=404, detail="Produto não encontrado")

    if data.tipo == "entrada":
        produto.estoque += data.quantidade
    elif data.tipo == "saida":
        produto.estoque -= data.quantidade
    elif data.tipo == "ajuste":
        produto.estoque = data.quantidade

    mov = MovimentoEstoque(
        produto_id=data.produto_id,
        tipo=data.tipo,
        quantidade=data.quantidade,
        motivo=data.motivo,
    )
    db.add(mov)
    db.commit()
    db.refresh(mov)

    nome = f"{produto.nome} - {produto.marca}" if produto.marca else produto.nome
    return {
        "id": mov.id,
        "produto_id": mov.produto_id,
        "produto_nome": nome,
        "tipo": mov.tipo,
        "quantidade": mov.quantidade,
        "motivo": mov.motivo,
        "criado_em": mov.criado_em.isoformat(),
    }


@router.get("/movimentos", response_model=List[MovimentoEstoqueOut])
def listar_movimentos(produto_id: Optional[int] = None, limit: int = 100, db: Session = Depends(get_db)):
    query = db.query(MovimentoEstoque)
    if produto_id:
        query = query.filter(MovimentoEstoque.produto_id == produto_id)
    movimentos = query.order_by(MovimentoEstoque.criado_em.desc()).limit(limit).all()
    result = []
    for m in movimentos:
        nome = None
        if m.produto:
            nome = f"{m.produto.nome} - {m.produto.marca}" if m.produto.marca else m.produto.nome
        result.append({
            "id": m.id,
            "produto_id": m.produto_id,
            "produto_nome": nome,
            "tipo": m.tipo,
            "quantidade": m.quantidade,
            "motivo": m.motivo,
            "criado_em": m.criado_em.isoformat(),
        })
    return result

File: artifacts/api-server/src/routes/estoque.ts
import { Router, type IRouter } from "express";
import { db } from "@workspace/db";
import { produtosTable, movimentosEstoqueTable } from "@workspace/db/schema";
import { eq, desc } from "drizzle-orm";
import { MovimentarEstoqueBody } from "@workspace/api-zod";

const router: IRouter = Router();

router.post("/movimento", async (req, res) => {
  const data = MovimentarEstoqueBody.parse(req.body);

  const [produto] = await db
    .select()
    .from(produtosTable)
    .where(eq(produtosTable.id, data.produto_id));

  if (!produto) {
    res.status(404).json({ ok: false, message: "Produto não encontrado" });
    return;
  }

  let novoEstoque = produto.estoque;
  if (data.tipo === "entrada") novoEstoque += data.quantidade;
  else if (data.tipo === "saida") novoEstoque -= data.quantidade;
  else if (data.tipo === "ajuste") novoEstoque = data.quantidade;

  await db
    .update(produtosTable)
    .set({ estoque: novoEstoque })
    .where(eq(produtosTable.id, data.produto_id));

  const [movimento] = await db
    .insert(movimentosEstoqueTable)
    .values({
      produto_id: data.produto_id,
      tipo: data.tipo,
      quantidade: data.quantidade,
      motivo: data.motivo ?? null,
    })
    .returning();

  res.status(201).json({
    ...movimento,
    produto_nome: produto.marca ? `${produto.nome} - ${produto.marca}` : produto.nome,
    criado_em: movimento.criado_em.toISOString(),
  });
});

router.get("/movimentos", async (req, res) => {
  const { produto_id, limit } = req.query as { produto_id?: string; limit?: string };
  const lim = parseInt(limit ?? "100") || 100;

  let query = db
    .select({
      movimento: movimentosEstoqueTable,
      produto_nome: produtosTable.nome,
      produto_marca: produtosTable.marca,
    })
    .from(movimentosEstoqueTable)
    .leftJoin(produtosTable, eq(movimentosEstoqueTable.produto_id, produtosTable.id))
    .orderBy(desc(movimentosEstoqueTable.criado_em))
    .limit(lim);

  if (produto_id) {
    query = query.where(eq(movimentosEstoqueTable.produto_id, parseInt(produto_id))) as typeof query;
  }

  const movimentos = await query;

  res.json(
    movimentos.map(({ movimento, produto_nome, produto_marca }) => ({
      ...movimento,
      produto_nome: produto_nome
        ? produto_marca
          ? `${produto_nome} - ${produto_marca}`
          : produto_nome
        : null,
      criado_em: movimento.criado_em.toISOString(),
    }))
  );
});

export default router;
